core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -262,36 +262,39 @@ private void advanceToNextPage() {
       // reference to the page to free and free it after releasing the lock of `MapIterator`.
       MemoryBlock pageToFree = null;
 
-      synchronized (this) {
-        int nextIdx = dataPages.indexOf(currentPage) + 1;
-        if (destructive && currentPage != null) {
-          dataPages.remove(currentPage);
-          pageToFree = currentPage;
-          nextIdx --;
-        }
-        if (dataPages.size() > nextIdx) {
-          currentPage = dataPages.get(nextIdx);
-          pageBaseObject = currentPage.getBaseObject();
-          offsetInPage = currentPage.getBaseOffset();
-          recordsInPage = UnsafeAlignedOffset.getSize(pageBaseObject, offsetInPage);
-          offsetInPage += UnsafeAlignedOffset.getUaoSize();
-        } else {
-          currentPage = null;
-          if (reader != null) {
-            handleFailedDelete();
+      try {
+        synchronized (this) {
+          int nextIdx = dataPages.indexOf(currentPage) + 1;
+          if (destructive && currentPage != null) {
+            dataPages.remove(currentPage);
+            pageToFree = currentPage;
+            nextIdx--;
           }
-          try {
-            Closeables.close(reader, /* swallowIOException = */ false);
```

Comment:
shall we close the reader outside of the `synchronized` block? Then we don't need to extra `try catch`.

Reply:
In the second `try`, it accesses `spillWriters`:
```
reader = spillWriters.getFirst().getReader(serializerManager)
```

The `synchronized` `spill` will change `spillWriters`. Seems to be unsafe if we close the reader and update `reader` like above outside `synchronized`?

Reply:
I'm not sure of the semantics here, but because the close block is changing several fields at once, it seems more conservative to leave them as they are within the synchronized block. At least, that is a separate issue.

Reply:
@cloud-fan What do you think?

Reply:
I think it's okay to go ahead as is